pipeline/runtime/run.go

```diff
@@ -110,12 +117,48 @@ func executeRunStep(ctx context.Context, f RunFunc, r *api.StartStepRequest, out
 					})
 				}
 			}
-			return exited, outputs, exportEnvs, artifact, outputsV2, string(optimizationState), err
 		} else if len(r.OutputVars) > 0 {
 			// only return err when output vars are expected
-			return exited, outputs, exportEnvs, artifact, nil, string(optimizationState), err
+			finalErr = err
+			for _, key := range r.OutputVars {
+				if _, ok := outputs[key]; ok {
+					output := &api.OutputV2{
+						Key:   key,
+						Value: outputs[key],
+						Type:  outputVariableTypeString,
+					}
+					outputsV2 = append(outputsV2, output)
+				}
+			}
+		} else {
```

Comment:
Why this else cond is needed ?

Reply:
when variables are exported from plugin . both the conditions will not satisfy . so the flow will need to go to else case to read them 
```
if len(r.Outputs) > 0 
len(r.OutputVars) > 0 
```